{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7193aee9-0304-4002-bc86-6ba7c9d08220",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------------------------\n",
      "\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "Loading required package: permute\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "This is vegan 2.6-4\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.16.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "! pheatmap() has been masked by ComplexHeatmap::pheatmap(). Most of the arguments\n",
      "   in the original pheatmap() are identically supported in the new function. You \n",
      "   can still use the original function by explicitly calling pheatmap::pheatmap().\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ComplexHeatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:pheatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.3     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32marrange()\u001b[39m   masks \u001b[34mdplyr\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m  masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mcount()\u001b[39m     masks \u001b[34mdplyr\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mdesc()\u001b[39m      masks \u001b[34mdplyr\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m  masks \u001b[34mdplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m   masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mid()\u001b[39m        masks \u001b[34mdplyr\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m      masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mmutate()\u001b[39m    masks \u001b[34mdplyr\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mrename()\u001b[39m    masks \u001b[34mdplyr\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32msummarise()\u001b[39m masks \u001b[34mdplyr\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32msummarize()\u001b[39m masks \u001b[34mdplyr\u001b[39m::summarize()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "suppressPackageStartupMessages(library(readxl))\n",
    "suppressPackageStartupMessages(library(dplyr))\n",
    "suppressPackageStartupMessages(library(tidyr))\n",
    "suppressPackageStartupMessages(library(tibble))\n",
    "suppressPackageStartupMessages(library(patchwork))\n",
    "suppressPackageStartupMessages(library(pheatmap))\n",
    "suppressPackageStartupMessages(library(stringr))\n",
    "library(hise)\n",
    "library(plyr)\n",
    "library(purrr)\n",
    "library(vegan)\n",
    "#library(arrow)\n",
    "library(parallel)\n",
    "library(ComplexHeatmap)\n",
    "library(ggsci)\n",
    "library(tidyverse)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "09ce8b62-3b4a-4f03-9f60-a2c9f29a8bad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m\u001b[90m[conflicted]\u001b[39m Will prefer \u001b[1m\u001b[34mdplyr\u001b[39m\u001b[22m::`%>%` over any other package.\n"
     ]
    }
   ],
   "source": [
    "library(stringr)\n",
    "library(readr)\n",
    "library(conflicted)\n",
    "library(sets)\n",
    "library(dplyr)\n",
    "conflict_prefer(\"%>%\", \"dplyr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e9c104d-fbc1-431a-a060-a86693c52f92",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load required libraries\n",
    "\n",
    "#library(sets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dbeaf116-07e9-4be1-8c09-c54a779e9052",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#install.packages(\"sets\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cb58fe9-2748-450b-9e5e-6e49ce871ea5",
   "metadata": {},
   "source": [
    "# Jaccard function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8f8a38f-b7bd-4172-8b25-e19da9a7f253",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# #Function to calculate Jaccard index\n",
    "# jaccard_index <- function(set1, set2) {\n",
    "#   intersection <- intersect(set1, set2)\n",
    "#   union <- union(set1, set2)\n",
    "#   return(length(intersection) / length(union))\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0d93777d-1b27-4522-bc60-f90fd48dbda1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "jaccard <- function(a, b) {\n",
    "    intersection = length(intersect(a, b))\n",
    "    union = length(a) + length(b) - intersection\n",
    "    return (intersection/union)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "27b03e73-16ee-4c6b-9b14-9e246a5a8581",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# list_to_set <- function(x) {\n",
    "#   unique(unlist(x))\n",
    "# }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f632d57e-b7d4-44ad-929f-18271224cc0c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# JI for each celltype (Paj<=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f96d8f4b-7242-4fcf-af73-95cd82e57652",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "master <- read.csv(\"/home/jupyter/IH-A-Aging-Analysis-Notebooks_old/Mansi_Notebooks/scRNA/CertPro_analysis/scripts/Figure_Notebooks/Intermediate_files/master_combined_DEG_list_all_celltypes_forfigs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1f374e6b-c711-4e0f-bc49-0dcaedfe4e4e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th><th scope=col>contrast</th><th scope=col>Direction</th><th scope=col>gene</th><th scope=col>celltype</th><th scope=col>sex</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>1.4553376</td><td>-1.58887408</td><td>0.3911621</td><td>-4.0619325</td><td>4.866815e-05</td><td>0.3688559</td><td>cohort.cohortGuid</td><td>BR1</td><td>CCDC174   </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>0.3984625</td><td> 0.41653866</td><td>0.5699233</td><td> 0.7308679</td><td>4.648598e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR2</td><td>AL669831.5</td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>3</td><td>1.6559504</td><td> 0.25216783</td><td>0.3641720</td><td> 0.6924415</td><td>4.886601e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR2</td><td>NOC2L     </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>4</td><td>1.2866563</td><td>-0.97078515</td><td>0.5119361</td><td>-1.8963013</td><td>5.792021e-02</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>ISG15     </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>5</td><td>2.7154615</td><td>-0.05010603</td><td>0.2936791</td><td>-0.1706149</td><td>8.645266e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>SDF4      </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>6</td><td>1.1378577</td><td>-0.29946048</td><td>0.4050389</td><td>-0.7393377</td><td>4.597020e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>B3GALT6   </td><td>ASDC</td><td>Female</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & X & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj & contrast & Direction & gene & celltype & sex\\\\\n",
       "  & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 1.4553376 & -1.58887408 & 0.3911621 & -4.0619325 & 4.866815e-05 & 0.3688559 & cohort.cohortGuid & BR1 & CCDC174    & ASDC & Female\\\\\n",
       "\t2 & 2 & 0.3984625 &  0.41653866 & 0.5699233 &  0.7308679 & 4.648598e-01 & 0.9968973 & cohort.cohortGuid & BR2 & AL669831.5 & ASDC & Female\\\\\n",
       "\t3 & 3 & 1.6559504 &  0.25216783 & 0.3641720 &  0.6924415 & 4.886601e-01 & 0.9968973 & cohort.cohortGuid & BR2 & NOC2L      & ASDC & Female\\\\\n",
       "\t4 & 4 & 1.2866563 & -0.97078515 & 0.5119361 & -1.8963013 & 5.792021e-02 & 0.9968973 & cohort.cohortGuid & BR1 & ISG15      & ASDC & Female\\\\\n",
       "\t5 & 5 & 2.7154615 & -0.05010603 & 0.2936791 & -0.1706149 & 8.645266e-01 & 0.9968973 & cohort.cohortGuid & BR1 & SDF4       & ASDC & Female\\\\\n",
       "\t6 & 6 & 1.1378577 & -0.29946048 & 0.4050389 & -0.7393377 & 4.597020e-01 & 0.9968973 & cohort.cohortGuid & BR1 & B3GALT6    & ASDC & Female\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | stat &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; | contrast &lt;chr&gt; | Direction &lt;chr&gt; | gene &lt;chr&gt; | celltype &lt;chr&gt; | sex &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 1.4553376 | -1.58887408 | 0.3911621 | -4.0619325 | 4.866815e-05 | 0.3688559 | cohort.cohortGuid | BR1 | CCDC174    | ASDC | Female |\n",
       "| 2 | 2 | 0.3984625 |  0.41653866 | 0.5699233 |  0.7308679 | 4.648598e-01 | 0.9968973 | cohort.cohortGuid | BR2 | AL669831.5 | ASDC | Female |\n",
       "| 3 | 3 | 1.6559504 |  0.25216783 | 0.3641720 |  0.6924415 | 4.886601e-01 | 0.9968973 | cohort.cohortGuid | BR2 | NOC2L      | ASDC | Female |\n",
       "| 4 | 4 | 1.2866563 | -0.97078515 | 0.5119361 | -1.8963013 | 5.792021e-02 | 0.9968973 | cohort.cohortGuid | BR1 | ISG15      | ASDC | Female |\n",
       "| 5 | 5 | 2.7154615 | -0.05010603 | 0.2936791 | -0.1706149 | 8.645266e-01 | 0.9968973 | cohort.cohortGuid | BR1 | SDF4       | ASDC | Female |\n",
       "| 6 | 6 | 1.1378577 | -0.29946048 | 0.4050389 | -0.7393377 | 4.597020e-01 | 0.9968973 | cohort.cohortGuid | BR1 | B3GALT6    | ASDC | Female |\n",
       "\n"
      ],
      "text/plain": [
       "  X baseMean  log2FoldChange lfcSE     stat       pvalue       padj     \n",
       "1 1 1.4553376 -1.58887408    0.3911621 -4.0619325 4.866815e-05 0.3688559\n",
       "2 2 0.3984625  0.41653866    0.5699233  0.7308679 4.648598e-01 0.9968973\n",
       "3 3 1.6559504  0.25216783    0.3641720  0.6924415 4.886601e-01 0.9968973\n",
       "4 4 1.2866563 -0.97078515    0.5119361 -1.8963013 5.792021e-02 0.9968973\n",
       "5 5 2.7154615 -0.05010603    0.2936791 -0.1706149 8.645266e-01 0.9968973\n",
       "6 6 1.1378577 -0.29946048    0.4050389 -0.7393377 4.597020e-01 0.9968973\n",
       "  contrast          Direction gene       celltype sex   \n",
       "1 cohort.cohortGuid BR1       CCDC174    ASDC     Female\n",
       "2 cohort.cohortGuid BR2       AL669831.5 ASDC     Female\n",
       "3 cohort.cohortGuid BR2       NOC2L      ASDC     Female\n",
       "4 cohort.cohortGuid BR1       ISG15      ASDC     Female\n",
       "5 cohort.cohortGuid BR1       SDF4       ASDC     Female\n",
       "6 cohort.cohortGuid BR1       B3GALT6    ASDC     Female"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(master)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1c020845-b2df-40a5-aca9-713e9eeb0ae6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ASDC'</li><li>'Activated memory B cell'</li><li>'Adaptive NK cell'</li><li>'BaEoMaP cell'</li><li>'C1Q+ CD16 monocyte'</li><li>'CD4 MAIT'</li><li>'CD8 MAIT'</li><li>'CD8aa'</li><li>'CD14+ cDC2'</li><li>'CD27+ effector B cell'</li><li>'CD27- effector B cell'</li><li>'CD56bright NK cell'</li><li>'CD95 memory B cell'</li><li>'CLP cell'</li><li>'CM CD4 T cell'</li><li>'CM CD8 T cell'</li><li>'CMP cell'</li><li>'Core CD14 monocyte'</li><li>'Core CD16 monocyte'</li><li>'Core memory B cell'</li><li>'Core naive B cell'</li><li>'Core naive CD4 T cell'</li><li>'Core naive CD8 T cell'</li><li>'DN T cell'</li><li>'Early memory B cell'</li><li>'Erythrocyte'</li><li>'GZMB+ Vd2 gdT'</li><li>'GZMB- CD27+ EM CD4 T cell'</li><li>'GZMB- CD27- EM CD4 T cell'</li><li>'GZMK+ CD27+ EM CD8 T cell'</li><li>'GZMK+ CD56dim NK cell'</li><li>'GZMK+ Vd2 gdT'</li><li>'GZMK+ memory CD4 Treg'</li><li>'GZMK- CD27+ EM CD8 T cell'</li><li>'GZMK- CD56dim NK cell'</li><li>'HLA-DRhi cDC2'</li><li>'IL1B+ CD14 monocyte'</li><li>'ILC'</li><li>'ISG+ CD14 monocyte'</li><li>'ISG+ CD16 monocyte'</li><li>'ISG+ CD56dim NK cell'</li><li>'ISG+ MAIT'</li><li>'ISG+ cDC2'</li><li>'ISG+ memory CD4 T cell'</li><li>'ISG+ memory CD8 T cell'</li><li>'ISG+ naive B cell'</li><li>'ISG+ naive CD4 T cell'</li><li>'ISG+ naive CD8 T cell'</li><li>'Intermediate monocyte'</li><li>'KLRB1+ memory CD4 Treg'</li><li>'KLRB1+ memory CD8 Treg'</li><li>'KLRF1+ GZMB+ CD27- EM CD8 T cell'</li><li>'KLRF1+ effector Vd1 gdT'</li><li>'KLRF1- GZMB+ CD27- EM CD8 T cell'</li><li>'KLRF1- GZMB+ CD27- memory CD4 T cell'</li><li>'KLRF1- effector Vd1 gdT'</li><li>'Memory CD4 Treg'</li><li>'Memory CD8 Treg'</li><li>'Naive CD4 Treg'</li><li>'Naive Vd1 gdT'</li><li>'Plasma cell'</li><li>'Platelet'</li><li>'Proliferating NK cell'</li><li>'Proliferating T cell'</li><li>'SOX4+ Vd1 gdT'</li><li>'SOX4+ naive CD4 T cell'</li><li>'SOX4+ naive CD8 T cell'</li><li>'Transitional B cell'</li><li>'Type 2 polarized memory B cell'</li><li>'cDC1'</li><li>'pDC'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ASDC'\n",
       "\\item 'Activated memory B cell'\n",
       "\\item 'Adaptive NK cell'\n",
       "\\item 'BaEoMaP cell'\n",
       "\\item 'C1Q+ CD16 monocyte'\n",
       "\\item 'CD4 MAIT'\n",
       "\\item 'CD8 MAIT'\n",
       "\\item 'CD8aa'\n",
       "\\item 'CD14+ cDC2'\n",
       "\\item 'CD27+ effector B cell'\n",
       "\\item 'CD27- effector B cell'\n",
       "\\item 'CD56bright NK cell'\n",
       "\\item 'CD95 memory B cell'\n",
       "\\item 'CLP cell'\n",
       "\\item 'CM CD4 T cell'\n",
       "\\item 'CM CD8 T cell'\n",
       "\\item 'CMP cell'\n",
       "\\item 'Core CD14 monocyte'\n",
       "\\item 'Core CD16 monocyte'\n",
       "\\item 'Core memory B cell'\n",
       "\\item 'Core naive B cell'\n",
       "\\item 'Core naive CD4 T cell'\n",
       "\\item 'Core naive CD8 T cell'\n",
       "\\item 'DN T cell'\n",
       "\\item 'Early memory B cell'\n",
       "\\item 'Erythrocyte'\n",
       "\\item 'GZMB+ Vd2 gdT'\n",
       "\\item 'GZMB- CD27+ EM CD4 T cell'\n",
       "\\item 'GZMB- CD27- EM CD4 T cell'\n",
       "\\item 'GZMK+ CD27+ EM CD8 T cell'\n",
       "\\item 'GZMK+ CD56dim NK cell'\n",
       "\\item 'GZMK+ Vd2 gdT'\n",
       "\\item 'GZMK+ memory CD4 Treg'\n",
       "\\item 'GZMK- CD27+ EM CD8 T cell'\n",
       "\\item 'GZMK- CD56dim NK cell'\n",
       "\\item 'HLA-DRhi cDC2'\n",
       "\\item 'IL1B+ CD14 monocyte'\n",
       "\\item 'ILC'\n",
       "\\item 'ISG+ CD14 monocyte'\n",
       "\\item 'ISG+ CD16 monocyte'\n",
       "\\item 'ISG+ CD56dim NK cell'\n",
       "\\item 'ISG+ MAIT'\n",
       "\\item 'ISG+ cDC2'\n",
       "\\item 'ISG+ memory CD4 T cell'\n",
       "\\item 'ISG+ memory CD8 T cell'\n",
       "\\item 'ISG+ naive B cell'\n",
       "\\item 'ISG+ naive CD4 T cell'\n",
       "\\item 'ISG+ naive CD8 T cell'\n",
       "\\item 'Intermediate monocyte'\n",
       "\\item 'KLRB1+ memory CD4 Treg'\n",
       "\\item 'KLRB1+ memory CD8 Treg'\n",
       "\\item 'KLRF1+ GZMB+ CD27- EM CD8 T cell'\n",
       "\\item 'KLRF1+ effector Vd1 gdT'\n",
       "\\item 'KLRF1- GZMB+ CD27- EM CD8 T cell'\n",
       "\\item 'KLRF1- GZMB+ CD27- memory CD4 T cell'\n",
       "\\item 'KLRF1- effector Vd1 gdT'\n",
       "\\item 'Memory CD4 Treg'\n",
       "\\item 'Memory CD8 Treg'\n",
       "\\item 'Naive CD4 Treg'\n",
       "\\item 'Naive Vd1 gdT'\n",
       "\\item 'Plasma cell'\n",
       "\\item 'Platelet'\n",
       "\\item 'Proliferating NK cell'\n",
       "\\item 'Proliferating T cell'\n",
       "\\item 'SOX4+ Vd1 gdT'\n",
       "\\item 'SOX4+ naive CD4 T cell'\n",
       "\\item 'SOX4+ naive CD8 T cell'\n",
       "\\item 'Transitional B cell'\n",
       "\\item 'Type 2 polarized memory B cell'\n",
       "\\item 'cDC1'\n",
       "\\item 'pDC'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ASDC'\n",
       "2. 'Activated memory B cell'\n",
       "3. 'Adaptive NK cell'\n",
       "4. 'BaEoMaP cell'\n",
       "5. 'C1Q+ CD16 monocyte'\n",
       "6. 'CD4 MAIT'\n",
       "7. 'CD8 MAIT'\n",
       "8. 'CD8aa'\n",
       "9. 'CD14+ cDC2'\n",
       "10. 'CD27+ effector B cell'\n",
       "11. 'CD27- effector B cell'\n",
       "12. 'CD56bright NK cell'\n",
       "13. 'CD95 memory B cell'\n",
       "14. 'CLP cell'\n",
       "15. 'CM CD4 T cell'\n",
       "16. 'CM CD8 T cell'\n",
       "17. 'CMP cell'\n",
       "18. 'Core CD14 monocyte'\n",
       "19. 'Core CD16 monocyte'\n",
       "20. 'Core memory B cell'\n",
       "21. 'Core naive B cell'\n",
       "22. 'Core naive CD4 T cell'\n",
       "23. 'Core naive CD8 T cell'\n",
       "24. 'DN T cell'\n",
       "25. 'Early memory B cell'\n",
       "26. 'Erythrocyte'\n",
       "27. 'GZMB+ Vd2 gdT'\n",
       "28. 'GZMB- CD27+ EM CD4 T cell'\n",
       "29. 'GZMB- CD27- EM CD4 T cell'\n",
       "30. 'GZMK+ CD27+ EM CD8 T cell'\n",
       "31. 'GZMK+ CD56dim NK cell'\n",
       "32. 'GZMK+ Vd2 gdT'\n",
       "33. 'GZMK+ memory CD4 Treg'\n",
       "34. 'GZMK- CD27+ EM CD8 T cell'\n",
       "35. 'GZMK- CD56dim NK cell'\n",
       "36. 'HLA-DRhi cDC2'\n",
       "37. 'IL1B+ CD14 monocyte'\n",
       "38. 'ILC'\n",
       "39. 'ISG+ CD14 monocyte'\n",
       "40. 'ISG+ CD16 monocyte'\n",
       "41. 'ISG+ CD56dim NK cell'\n",
       "42. 'ISG+ MAIT'\n",
       "43. 'ISG+ cDC2'\n",
       "44. 'ISG+ memory CD4 T cell'\n",
       "45. 'ISG+ memory CD8 T cell'\n",
       "46. 'ISG+ naive B cell'\n",
       "47. 'ISG+ naive CD4 T cell'\n",
       "48. 'ISG+ naive CD8 T cell'\n",
       "49. 'Intermediate monocyte'\n",
       "50. 'KLRB1+ memory CD4 Treg'\n",
       "51. 'KLRB1+ memory CD8 Treg'\n",
       "52. 'KLRF1+ GZMB+ CD27- EM CD8 T cell'\n",
       "53. 'KLRF1+ effector Vd1 gdT'\n",
       "54. 'KLRF1- GZMB+ CD27- EM CD8 T cell'\n",
       "55. 'KLRF1- GZMB+ CD27- memory CD4 T cell'\n",
       "56. 'KLRF1- effector Vd1 gdT'\n",
       "57. 'Memory CD4 Treg'\n",
       "58. 'Memory CD8 Treg'\n",
       "59. 'Naive CD4 Treg'\n",
       "60. 'Naive Vd1 gdT'\n",
       "61. 'Plasma cell'\n",
       "62. 'Platelet'\n",
       "63. 'Proliferating NK cell'\n",
       "64. 'Proliferating T cell'\n",
       "65. 'SOX4+ Vd1 gdT'\n",
       "66. 'SOX4+ naive CD4 T cell'\n",
       "67. 'SOX4+ naive CD8 T cell'\n",
       "68. 'Transitional B cell'\n",
       "69. 'Type 2 polarized memory B cell'\n",
       "70. 'cDC1'\n",
       "71. 'pDC'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"ASDC\"                                \n",
       " [2] \"Activated memory B cell\"             \n",
       " [3] \"Adaptive NK cell\"                    \n",
       " [4] \"BaEoMaP cell\"                        \n",
       " [5] \"C1Q+ CD16 monocyte\"                  \n",
       " [6] \"CD4 MAIT\"                            \n",
       " [7] \"CD8 MAIT\"                            \n",
       " [8] \"CD8aa\"                               \n",
       " [9] \"CD14+ cDC2\"                          \n",
       "[10] \"CD27+ effector B cell\"               \n",
       "[11] \"CD27- effector B cell\"               \n",
       "[12] \"CD56bright NK cell\"                  \n",
       "[13] \"CD95 memory B cell\"                  \n",
       "[14] \"CLP cell\"                            \n",
       "[15] \"CM CD4 T cell\"                       \n",
       "[16] \"CM CD8 T cell\"                       \n",
       "[17] \"CMP cell\"                            \n",
       "[18] \"Core CD14 monocyte\"                  \n",
       "[19] \"Core CD16 monocyte\"                  \n",
       "[20] \"Core memory B cell\"                  \n",
       "[21] \"Core naive B cell\"                   \n",
       "[22] \"Core naive CD4 T cell\"               \n",
       "[23] \"Core naive CD8 T cell\"               \n",
       "[24] \"DN T cell\"                           \n",
       "[25] \"Early memory B cell\"                 \n",
       "[26] \"Erythrocyte\"                         \n",
       "[27] \"GZMB+ Vd2 gdT\"                       \n",
       "[28] \"GZMB- CD27+ EM CD4 T cell\"           \n",
       "[29] \"GZMB- CD27- EM CD4 T cell\"           \n",
       "[30] \"GZMK+ CD27+ EM CD8 T cell\"           \n",
       "[31] \"GZMK+ CD56dim NK cell\"               \n",
       "[32] \"GZMK+ Vd2 gdT\"                       \n",
       "[33] \"GZMK+ memory CD4 Treg\"               \n",
       "[34] \"GZMK- CD27+ EM CD8 T cell\"           \n",
       "[35] \"GZMK- CD56dim NK cell\"               \n",
       "[36] \"HLA-DRhi cDC2\"                       \n",
       "[37] \"IL1B+ CD14 monocyte\"                 \n",
       "[38] \"ILC\"                                 \n",
       "[39] \"ISG+ CD14 monocyte\"                  \n",
       "[40] \"ISG+ CD16 monocyte\"                  \n",
       "[41] \"ISG+ CD56dim NK cell\"                \n",
       "[42] \"ISG+ MAIT\"                           \n",
       "[43] \"ISG+ cDC2\"                           \n",
       "[44] \"ISG+ memory CD4 T cell\"              \n",
       "[45] \"ISG+ memory CD8 T cell\"              \n",
       "[46] \"ISG+ naive B cell\"                   \n",
       "[47] \"ISG+ naive CD4 T cell\"               \n",
       "[48] \"ISG+ naive CD8 T cell\"               \n",
       "[49] \"Intermediate monocyte\"               \n",
       "[50] \"KLRB1+ memory CD4 Treg\"              \n",
       "[51] \"KLRB1+ memory CD8 Treg\"              \n",
       "[52] \"KLRF1+ GZMB+ CD27- EM CD8 T cell\"    \n",
       "[53] \"KLRF1+ effector Vd1 gdT\"             \n",
       "[54] \"KLRF1- GZMB+ CD27- EM CD8 T cell\"    \n",
       "[55] \"KLRF1- GZMB+ CD27- memory CD4 T cell\"\n",
       "[56] \"KLRF1- effector Vd1 gdT\"             \n",
       "[57] \"Memory CD4 Treg\"                     \n",
       "[58] \"Memory CD8 Treg\"                     \n",
       "[59] \"Naive CD4 Treg\"                      \n",
       "[60] \"Naive Vd1 gdT\"                       \n",
       "[61] \"Plasma cell\"                         \n",
       "[62] \"Platelet\"                            \n",
       "[63] \"Proliferating NK cell\"               \n",
       "[64] \"Proliferating T cell\"                \n",
       "[65] \"SOX4+ Vd1 gdT\"                       \n",
       "[66] \"SOX4+ naive CD4 T cell\"              \n",
       "[67] \"SOX4+ naive CD8 T cell\"              \n",
       "[68] \"Transitional B cell\"                 \n",
       "[69] \"Type 2 polarized memory B cell\"      \n",
       "[70] \"cDC1\"                                \n",
       "[71] \"pDC\"                                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(master$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7f1ae50f-7702-4e94-bda7-cd02d1f5908e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th><th scope=col>contrast</th><th scope=col>Direction</th><th scope=col>gene</th><th scope=col>celltype</th><th scope=col>sex</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>1.4553376</td><td>-1.58887408</td><td>0.3911621</td><td>-4.0619325</td><td>4.866815e-05</td><td>0.3688559</td><td>cohort.cohortGuid</td><td>BR1</td><td>CCDC174   </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>0.3984625</td><td> 0.41653866</td><td>0.5699233</td><td> 0.7308679</td><td>4.648598e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR2</td><td>AL669831.5</td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>3</td><td>1.6559504</td><td> 0.25216783</td><td>0.3641720</td><td> 0.6924415</td><td>4.886601e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR2</td><td>NOC2L     </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>4</td><td>1.2866563</td><td>-0.97078515</td><td>0.5119361</td><td>-1.8963013</td><td>5.792021e-02</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>ISG15     </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>5</td><td>2.7154615</td><td>-0.05010603</td><td>0.2936791</td><td>-0.1706149</td><td>8.645266e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>SDF4      </td><td>ASDC</td><td>Female</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>6</td><td>1.1378577</td><td>-0.29946048</td><td>0.4050389</td><td>-0.7393377</td><td>4.597020e-01</td><td>0.9968973</td><td>cohort.cohortGuid</td><td>BR1</td><td>B3GALT6   </td><td>ASDC</td><td>Female</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & X & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj & contrast & Direction & gene & celltype & sex\\\\\n",
       "  & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 1.4553376 & -1.58887408 & 0.3911621 & -4.0619325 & 4.866815e-05 & 0.3688559 & cohort.cohortGuid & BR1 & CCDC174    & ASDC & Female\\\\\n",
       "\t2 & 2 & 0.3984625 &  0.41653866 & 0.5699233 &  0.7308679 & 4.648598e-01 & 0.9968973 & cohort.cohortGuid & BR2 & AL669831.5 & ASDC & Female\\\\\n",
       "\t3 & 3 & 1.6559504 &  0.25216783 & 0.3641720 &  0.6924415 & 4.886601e-01 & 0.9968973 & cohort.cohortGuid & BR2 & NOC2L      & ASDC & Female\\\\\n",
       "\t4 & 4 & 1.2866563 & -0.97078515 & 0.5119361 & -1.8963013 & 5.792021e-02 & 0.9968973 & cohort.cohortGuid & BR1 & ISG15      & ASDC & Female\\\\\n",
       "\t5 & 5 & 2.7154615 & -0.05010603 & 0.2936791 & -0.1706149 & 8.645266e-01 & 0.9968973 & cohort.cohortGuid & BR1 & SDF4       & ASDC & Female\\\\\n",
       "\t6 & 6 & 1.1378577 & -0.29946048 & 0.4050389 & -0.7393377 & 4.597020e-01 & 0.9968973 & cohort.cohortGuid & BR1 & B3GALT6    & ASDC & Female\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | stat &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; | contrast &lt;chr&gt; | Direction &lt;chr&gt; | gene &lt;chr&gt; | celltype &lt;chr&gt; | sex &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 1.4553376 | -1.58887408 | 0.3911621 | -4.0619325 | 4.866815e-05 | 0.3688559 | cohort.cohortGuid | BR1 | CCDC174    | ASDC | Female |\n",
       "| 2 | 2 | 0.3984625 |  0.41653866 | 0.5699233 |  0.7308679 | 4.648598e-01 | 0.9968973 | cohort.cohortGuid | BR2 | AL669831.5 | ASDC | Female |\n",
       "| 3 | 3 | 1.6559504 |  0.25216783 | 0.3641720 |  0.6924415 | 4.886601e-01 | 0.9968973 | cohort.cohortGuid | BR2 | NOC2L      | ASDC | Female |\n",
       "| 4 | 4 | 1.2866563 | -0.97078515 | 0.5119361 | -1.8963013 | 5.792021e-02 | 0.9968973 | cohort.cohortGuid | BR1 | ISG15      | ASDC | Female |\n",
       "| 5 | 5 | 2.7154615 | -0.05010603 | 0.2936791 | -0.1706149 | 8.645266e-01 | 0.9968973 | cohort.cohortGuid | BR1 | SDF4       | ASDC | Female |\n",
       "| 6 | 6 | 1.1378577 | -0.29946048 | 0.4050389 | -0.7393377 | 4.597020e-01 | 0.9968973 | cohort.cohortGuid | BR1 | B3GALT6    | ASDC | Female |\n",
       "\n"
      ],
      "text/plain": [
       "  X baseMean  log2FoldChange lfcSE     stat       pvalue       padj     \n",
       "1 1 1.4553376 -1.58887408    0.3911621 -4.0619325 4.866815e-05 0.3688559\n",
       "2 2 0.3984625  0.41653866    0.5699233  0.7308679 4.648598e-01 0.9968973\n",
       "3 3 1.6559504  0.25216783    0.3641720  0.6924415 4.886601e-01 0.9968973\n",
       "4 4 1.2866563 -0.97078515    0.5119361 -1.8963013 5.792021e-02 0.9968973\n",
       "5 5 2.7154615 -0.05010603    0.2936791 -0.1706149 8.645266e-01 0.9968973\n",
       "6 6 1.1378577 -0.29946048    0.4050389 -0.7393377 4.597020e-01 0.9968973\n",
       "  contrast          Direction gene       celltype sex   \n",
       "1 cohort.cohortGuid BR1       CCDC174    ASDC     Female\n",
       "2 cohort.cohortGuid BR2       AL669831.5 ASDC     Female\n",
       "3 cohort.cohortGuid BR2       NOC2L      ASDC     Female\n",
       "4 cohort.cohortGuid BR1       ISG15      ASDC     Female\n",
       "5 cohort.cohortGuid BR1       SDF4       ASDC     Female\n",
       "6 cohort.cohortGuid BR1       B3GALT6    ASDC     Female"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(master)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "360c8db0-3145-4720-acac-1980747182c8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               celltype jaccard_similarity total_DEGs_male\n",
      "1                                  ASDC                NaN               0\n",
      "2               Activated memory B cell                NaN               0\n",
      "3                      Adaptive NK cell                NaN               0\n",
      "4                          BaEoMaP cell                NaN               0\n",
      "5                    C1Q+ CD16 monocyte                NaN               0\n",
      "6                              CD4 MAIT                NaN               0\n",
      "7                              CD8 MAIT         0.00000000               0\n",
      "8                                 CD8aa         0.00000000               5\n",
      "9                            CD14+ cDC2                NaN               0\n",
      "10                CD27+ effector B cell         0.00000000               0\n",
      "11                CD27- effector B cell         0.00000000               1\n",
      "12                   CD56bright NK cell                NaN               0\n",
      "13                   CD95 memory B cell                NaN               0\n",
      "14                             CLP cell                NaN               0\n",
      "15                        CM CD4 T cell         0.07142857               8\n",
      "16                        CM CD8 T cell         0.04938272              68\n",
      "17                             CMP cell                NaN               0\n",
      "18                   Core CD14 monocyte         0.00000000               2\n",
      "19                   Core CD16 monocyte         0.00000000               3\n",
      "20                   Core memory B cell         0.10000000               5\n",
      "21                    Core naive B cell                NaN               0\n",
      "22                Core naive CD4 T cell         0.16783217             167\n",
      "23                Core naive CD8 T cell         0.15053763              40\n",
      "24                            DN T cell         0.00000000               0\n",
      "25                  Early memory B cell                NaN               0\n",
      "26                          Erythrocyte                NaN               0\n",
      "27                        GZMB+ Vd2 gdT         0.00000000               0\n",
      "28            GZMB- CD27+ EM CD4 T cell         0.04761905              10\n",
      "29            GZMB- CD27- EM CD4 T cell         0.11111111               3\n",
      "30            GZMK+ CD27+ EM CD8 T cell         0.00000000              12\n",
      "31                GZMK+ CD56dim NK cell                NaN               0\n",
      "32                        GZMK+ Vd2 gdT                NaN               0\n",
      "33                GZMK+ memory CD4 Treg                NaN               0\n",
      "34            GZMK- CD27+ EM CD8 T cell         0.00000000               0\n",
      "35                GZMK- CD56dim NK cell                NaN               0\n",
      "36                        HLA-DRhi cDC2         0.00000000               0\n",
      "37                  IL1B+ CD14 monocyte         0.00000000               0\n",
      "38                                  ILC                NaN               0\n",
      "39                   ISG+ CD14 monocyte         0.00000000               1\n",
      "40                   ISG+ CD16 monocyte         0.00000000               0\n",
      "41                 ISG+ CD56dim NK cell                NaN               0\n",
      "42                            ISG+ MAIT                NaN               0\n",
      "43                            ISG+ cDC2                NaN               0\n",
      "44               ISG+ memory CD4 T cell                NaN               0\n",
      "45               ISG+ memory CD8 T cell                NaN               0\n",
      "46                    ISG+ naive B cell                NaN               0\n",
      "47                ISG+ naive CD4 T cell         0.00000000               0\n",
      "48                ISG+ naive CD8 T cell                NaN               0\n",
      "49                Intermediate monocyte         0.00000000               3\n",
      "50               KLRB1+ memory CD4 Treg                NaN               0\n",
      "51               KLRB1+ memory CD8 Treg                NaN               0\n",
      "52     KLRF1+ GZMB+ CD27- EM CD8 T cell         0.00000000               1\n",
      "53              KLRF1+ effector Vd1 gdT                NaN               0\n",
      "54     KLRF1- GZMB+ CD27- EM CD8 T cell         0.00000000               0\n",
      "55 KLRF1- GZMB+ CD27- memory CD4 T cell         0.00000000               2\n",
      "56              KLRF1- effector Vd1 gdT                NaN               0\n",
      "57                      Memory CD4 Treg         0.00000000               3\n",
      "58                      Memory CD8 Treg                NaN               0\n",
      "59                       Naive CD4 Treg         0.04347826              24\n",
      "60                        Naive Vd1 gdT                NaN               0\n",
      "61                          Plasma cell         0.00000000               0\n",
      "62                             Platelet                NaN               0\n",
      "63                Proliferating NK cell                NaN               0\n",
      "64                 Proliferating T cell                NaN               0\n",
      "65                        SOX4+ Vd1 gdT                NaN               0\n",
      "66               SOX4+ naive CD4 T cell                NaN               0\n",
      "67               SOX4+ naive CD8 T cell                NaN               0\n",
      "68                  Transitional B cell                NaN               0\n",
      "69       Type 2 polarized memory B cell                NaN               0\n",
      "70                                 cDC1                NaN               0\n",
      "71                                  pDC                NaN               0\n",
      "   total_DEGs_female common_DEGs\n",
      "1                  0           0\n",
      "2                  0           0\n",
      "3                  0           0\n",
      "4                  0           0\n",
      "5                  0           0\n",
      "6                  0           0\n",
      "7                  4           0\n",
      "8                 10           0\n",
      "9                  0           0\n",
      "10                 3           0\n",
      "11                 2           0\n",
      "12                 0           0\n",
      "13                 0           0\n",
      "14                 0           0\n",
      "15                 7           1\n",
      "16                17           4\n",
      "17                 0           0\n",
      "18                25           0\n",
      "19                 0           0\n",
      "20                 6           1\n",
      "21                 0           0\n",
      "22               167          48\n",
      "23                67          14\n",
      "24                 3           0\n",
      "25                 0           0\n",
      "26                 0           0\n",
      "27                13           0\n",
      "28                12           1\n",
      "29                 7           1\n",
      "30                 3           0\n",
      "31                 0           0\n",
      "32                 0           0\n",
      "33                 0           0\n",
      "34                 2           0\n",
      "35                 0           0\n",
      "36                 1           0\n",
      "37                 4           0\n",
      "38                 0           0\n",
      "39                 4           0\n",
      "40                 1           0\n",
      "41                 0           0\n",
      "42                 0           0\n",
      "43                 0           0\n",
      "44                 0           0\n",
      "45                 0           0\n",
      "46                 0           0\n",
      "47                 7           0\n",
      "48                 0           0\n",
      "49                 0           0\n",
      "50                 0           0\n",
      "51                 0           0\n",
      "52                 2           0\n",
      "53                 0           0\n",
      "54                 1           0\n",
      "55                 0           0\n",
      "56                 0           0\n",
      "57                 0           0\n",
      "58                 0           0\n",
      "59                24           2\n",
      "60                 0           0\n",
      "61                 2           0\n",
      "62                 0           0\n",
      "63                 0           0\n",
      "64                 0           0\n",
      "65                 0           0\n",
      "66                 0           0\n",
      "67                 0           0\n",
      "68                 0           0\n",
      "69                 0           0\n",
      "70                 0           0\n",
      "71                 0           0\n"
     ]
    }
   ],
   "source": [
    "# library(sets)\n",
    "# library(dplyr)\n",
    "\n",
    "# Initialize an empty dataframe to store results\n",
    "results <- data.frame(celltype = character(),\n",
    "                      jaccard_similarity = numeric(),\n",
    "                       total_DEGs_male = integer(),\n",
    "                       total_DEGs_female = integer(),\n",
    "                       common_DEGs = integer(),\n",
    "                       stringsAsFactors = FALSE)\n",
    "\n",
    "# Iterate through cell types\n",
    "for (cell in unique(master$celltype)) {\n",
    "  \n",
    "    # Filter data for the current cell type\n",
    "  cell_filt <- master %>% dplyr::filter(celltype == cell, padj <= 0.05)\n",
    "  \n",
    "  # Filter DEGs for males and females\n",
    "  DEG_male <- cell_filt[cell_filt$padj <= 0.1 & cell_filt$sex == \"Male\", \"gene\"]\n",
    "  DEG_female <- cell_filt[cell_filt$padj <= 0.1 & cell_filt$sex == \"Female\", \"gene\"]\n",
    "  \n",
    "  # Remove NA values from DEGs\n",
    "  DEG_male <- na.omit(DEG_male)\n",
    "  DEG_female <- na.omit(DEG_female)\n",
    "  \n",
    "  # Convert filtered DEGs to sets\n",
    "  set_DEG_male <- as.set(DEG_male)\n",
    "  set_DEG_female <- as.set(DEG_female)\n",
    "  \n",
    "  # Calculate total DEGs counts\n",
    "  total_DEGs_male <- length(DEG_male)\n",
    "  total_DEGs_female <- length(DEG_female)\n",
    "  \n",
    "  # Calculate common DEGs counts\n",
    "  common_DEGs <- length(intersect(set_DEG_male, set_DEG_female))\n",
    "  \n",
    "  # Calculate Jaccard similarity\n",
    "  jaccard_sim <- jaccard(set_DEG_male, set_DEG_female)\n",
    "  \n",
    "  # Append results to dataframe\n",
    "  results <- rbind(results, data.frame(celltype = cell,\n",
    "                                       jaccard_similarity = jaccard_sim,\n",
    "                                        total_DEGs_male = total_DEGs_male,\n",
    "                                        total_DEGs_female = total_DEGs_female,\n",
    "                                        common_DEGs = common_DEGs))\n",
    "}\n",
    "\n",
    "# Print results\n",
    "print(results)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "168bcea1-5fe3-4a28-8abf-f0ce2b14525c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "write.csv(results, \"/home/jupyter/IH-A-Aging-Analysis-Notebooks_old/Mansi_Notebooks/scRNA/CertPro_analysis/scripts/Figure_Notebooks/Intermediate_files/Jaccard_index/Jaccard_index_df_FvsM_each_celltype_padj0.05_v2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a940d6-5342-43b3-853d-332b45695bbe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2db70f90-a30c-43ea-8396-5c1d6fe9f3ef",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# JI for each celltype (Paj<=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8d0dbb28-f439-40da-9a0c-8bb25d72b0c1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "master <- read.csv(\"/home/jupyter/IH-A-Aging-Analysis-Notebooks_old/Mansi_Notebooks/scRNA/Figure_Notebooks/Intermediate_files/master_combined_DEG_list_all_celltypes_forfigs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b7199df2-2dfc-45da-8e8b-dcafa8d16bfe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th><th scope=col>celltype</th><th scope=col>Direction</th><th scope=col>Year</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NOC2L  </td><td>0.4682601</td><td>-0.33564453</td><td>0.8327112</td><td>-0.40307434</td><td>0.6868935</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR1</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ISG15  </td><td>0.7716756</td><td> 0.03101969</td><td>0.5329482</td><td> 0.05820395</td><td>0.9535862</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR2</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>SDF4   </td><td>0.5631814</td><td> 0.01246954</td><td>0.6510630</td><td> 0.01915259</td><td>0.9847194</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR2</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>B3GALT6</td><td>0.3151741</td><td> 0.61845489</td><td>0.8779896</td><td> 0.70439886</td><td>0.4811844</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR2</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>UBE2J2 </td><td>0.4517144</td><td> 0.17761642</td><td>0.6528145</td><td> 0.27207793</td><td>0.7855621</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR2</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>INTS11 </td><td>0.3656259</td><td> 0.07206422</td><td>0.6856389</td><td> 0.10510521</td><td>0.9162923</td><td>0.9998266</td><td>Activated memory B cell</td><td>BR2</td><td>Y1</td><td>DEGs_Male</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & X & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj & celltype & Direction & Year & type\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & NOC2L   & 0.4682601 & -0.33564453 & 0.8327112 & -0.40307434 & 0.6868935 & 0.9998266 & Activated memory B cell & BR1 & Y1 & DEGs\\_Male\\\\\n",
       "\t2 & ISG15   & 0.7716756 &  0.03101969 & 0.5329482 &  0.05820395 & 0.9535862 & 0.9998266 & Activated memory B cell & BR2 & Y1 & DEGs\\_Male\\\\\n",
       "\t3 & SDF4    & 0.5631814 &  0.01246954 & 0.6510630 &  0.01915259 & 0.9847194 & 0.9998266 & Activated memory B cell & BR2 & Y1 & DEGs\\_Male\\\\\n",
       "\t4 & B3GALT6 & 0.3151741 &  0.61845489 & 0.8779896 &  0.70439886 & 0.4811844 & 0.9998266 & Activated memory B cell & BR2 & Y1 & DEGs\\_Male\\\\\n",
       "\t5 & UBE2J2  & 0.4517144 &  0.17761642 & 0.6528145 &  0.27207793 & 0.7855621 & 0.9998266 & Activated memory B cell & BR2 & Y1 & DEGs\\_Male\\\\\n",
       "\t6 & INTS11  & 0.3656259 &  0.07206422 & 0.6856389 &  0.10510521 & 0.9162923 & 0.9998266 & Activated memory B cell & BR2 & Y1 & DEGs\\_Male\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 11\n",
       "\n",
       "| <!--/--> | X &lt;chr&gt; | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | stat &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; | celltype &lt;chr&gt; | Direction &lt;chr&gt; | Year &lt;chr&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NOC2L   | 0.4682601 | -0.33564453 | 0.8327112 | -0.40307434 | 0.6868935 | 0.9998266 | Activated memory B cell | BR1 | Y1 | DEGs_Male |\n",
       "| 2 | ISG15   | 0.7716756 |  0.03101969 | 0.5329482 |  0.05820395 | 0.9535862 | 0.9998266 | Activated memory B cell | BR2 | Y1 | DEGs_Male |\n",
       "| 3 | SDF4    | 0.5631814 |  0.01246954 | 0.6510630 |  0.01915259 | 0.9847194 | 0.9998266 | Activated memory B cell | BR2 | Y1 | DEGs_Male |\n",
       "| 4 | B3GALT6 | 0.3151741 |  0.61845489 | 0.8779896 |  0.70439886 | 0.4811844 | 0.9998266 | Activated memory B cell | BR2 | Y1 | DEGs_Male |\n",
       "| 5 | UBE2J2  | 0.4517144 |  0.17761642 | 0.6528145 |  0.27207793 | 0.7855621 | 0.9998266 | Activated memory B cell | BR2 | Y1 | DEGs_Male |\n",
       "| 6 | INTS11  | 0.3656259 |  0.07206422 | 0.6856389 |  0.10510521 | 0.9162923 | 0.9998266 | Activated memory B cell | BR2 | Y1 | DEGs_Male |\n",
       "\n"
      ],
      "text/plain": [
       "  X       baseMean  log2FoldChange lfcSE     stat        pvalue    padj     \n",
       "1 NOC2L   0.4682601 -0.33564453    0.8327112 -0.40307434 0.6868935 0.9998266\n",
       "2 ISG15   0.7716756  0.03101969    0.5329482  0.05820395 0.9535862 0.9998266\n",
       "3 SDF4    0.5631814  0.01246954    0.6510630  0.01915259 0.9847194 0.9998266\n",
       "4 B3GALT6 0.3151741  0.61845489    0.8779896  0.70439886 0.4811844 0.9998266\n",
       "5 UBE2J2  0.4517144  0.17761642    0.6528145  0.27207793 0.7855621 0.9998266\n",
       "6 INTS11  0.3656259  0.07206422    0.6856389  0.10510521 0.9162923 0.9998266\n",
       "  celltype                Direction Year type     \n",
       "1 Activated memory B cell BR1       Y1   DEGs_Male\n",
       "2 Activated memory B cell BR2       Y1   DEGs_Male\n",
       "3 Activated memory B cell BR2       Y1   DEGs_Male\n",
       "4 Activated memory B cell BR2       Y1   DEGs_Male\n",
       "5 Activated memory B cell BR2       Y1   DEGs_Male\n",
       "6 Activated memory B cell BR2       Y1   DEGs_Male"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(master)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2065e959-e9de-4341-b624-064779981df1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Activated memory B cell'</li><li>'Adaptive NK cell'</li><li>'ASDC'</li><li>'BaEoMaP cell'</li><li>'C1Q+ CD16 monocyte'</li><li>'CD14+ cDC2'</li><li>'CD27- effector B cell'</li><li>'CD27+ effector B cell'</li><li>'CD4 MAIT'</li><li>'CD56bright NK cell'</li><li>'CD8 MAIT'</li><li>'CD8aa'</li><li>'CD95 memory B cell'</li><li>'cDC1'</li><li>'CLP cell'</li><li>'CM CD4 T cell'</li><li>'CM CD8 T cell'</li><li>'CMP cell'</li><li>'Core CD14 monocyte'</li><li>'Core CD16 monocyte'</li><li>'Core memory B cell'</li><li>'Core naive B cell'</li><li>'Core naive CD4 T cell'</li><li>'Core naive CD8 T cell '</li><li>'DN T cell'</li><li>'Early memory B cell'</li><li>'Erythrocyte'</li><li>'GZMB- CD27- EM CD4 T cell'</li><li>'GZMB- CD27+ EM CD4 T cell'</li><li>'GZMB+ Vd2 gdT'</li><li>'GZMK- CD27+ EM CD8 T cell'</li><li>'GZMK- CD56dim NK cell'</li><li>'GZMK+ CD27+ EM CD8 T cell'</li><li>'GZMK+ CD56dim NK cell'</li><li>'GZMK+ memory CD4 Treg'</li><li>'GZMK+ Vd2 gdT'</li><li>'HLA-DRhi cDC2'</li><li>'IL1B+ CD14 monocyte'</li><li>'ILC'</li><li>'Intermediate monocyte'</li><li>'ISG+ CD14 monocyte'</li><li>'ISG+ CD16 monocyte'</li><li>'ISG+ CD56dim NK cell'</li><li>'ISG+ cDC2'</li><li>'ISG+ MAIT'</li><li>'ISG+ memory CD4 T cell'</li><li>'ISG+ memory CD8 T cell'</li><li>'ISG+ naive B cell'</li><li>'ISG+ naive CD4 T cell'</li><li>'ISG+ naive CD8 T cell'</li><li>'KLRB1+ memory CD4 Treg'</li><li>'KLRB1+ memory CD8 Treg'</li><li>'KLRF1- effector Vd1 gdT'</li><li>'KLRF1- GZMB+ CD27- EM CD8 T cell'</li><li>'KLRF1- GZMB+ CD27- memory CD4 T cell'</li><li>'KLRF1+ effector Vd1 gdT'</li><li>'KLRF1+ GZMB+ CD27- EM CD8 T cell'</li><li>'Memory CD4 Treg'</li><li>'Memory CD8 Treg'</li><li>'Naive CD4 Treg'</li><li>'Naive Vd1 gdT'</li><li>'pDC'</li><li>'Plasma cell'</li><li>'Platelet'</li><li>'Proliferating NK cell'</li><li>'Proliferating T cell'</li><li>'SOX4+ naive CD4 T cell'</li><li>'SOX4+ naive CD8 T cell'</li><li>'SOX4+ Vd1 gdT'</li><li>'Transitional B cell'</li><li>'Type 2 polarized memory B cell'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Activated memory B cell'\n",
       "\\item 'Adaptive NK cell'\n",
       "\\item 'ASDC'\n",
       "\\item 'BaEoMaP cell'\n",
       "\\item 'C1Q+ CD16 monocyte'\n",
       "\\item 'CD14+ cDC2'\n",
       "\\item 'CD27- effector B cell'\n",
       "\\item 'CD27+ effector B cell'\n",
       "\\item 'CD4 MAIT'\n",
       "\\item 'CD56bright NK cell'\n",
       "\\item 'CD8 MAIT'\n",
       "\\item 'CD8aa'\n",
       "\\item 'CD95 memory B cell'\n",
       "\\item 'cDC1'\n",
       "\\item 'CLP cell'\n",
       "\\item 'CM CD4 T cell'\n",
       "\\item 'CM CD8 T cell'\n",
       "\\item 'CMP cell'\n",
       "\\item 'Core CD14 monocyte'\n",
       "\\item 'Core CD16 monocyte'\n",
       "\\item 'Core memory B cell'\n",
       "\\item 'Core naive B cell'\n",
       "\\item 'Core naive CD4 T cell'\n",
       "\\item 'Core naive CD8 T cell '\n",
       "\\item 'DN T cell'\n",
       "\\item 'Early memory B cell'\n",
       "\\item 'Erythrocyte'\n",
       "\\item 'GZMB- CD27- EM CD4 T cell'\n",
       "\\item 'GZMB- CD27+ EM CD4 T cell'\n",
       "\\item 'GZMB+ Vd2 gdT'\n",
       "\\item 'GZMK- CD27+ EM CD8 T cell'\n",
       "\\item 'GZMK- CD56dim NK cell'\n",
       "\\item 'GZMK+ CD27+ EM CD8 T cell'\n",
       "\\item 'GZMK+ CD56dim NK cell'\n",
       "\\item 'GZMK+ memory CD4 Treg'\n",
       "\\item 'GZMK+ Vd2 gdT'\n",
       "\\item 'HLA-DRhi cDC2'\n",
       "\\item 'IL1B+ CD14 monocyte'\n",
       "\\item 'ILC'\n",
       "\\item 'Intermediate monocyte'\n",
       "\\item 'ISG+ CD14 monocyte'\n",
       "\\item 'ISG+ CD16 monocyte'\n",
       "\\item 'ISG+ CD56dim NK cell'\n",
       "\\item 'ISG+ cDC2'\n",
       "\\item 'ISG+ MAIT'\n",
       "\\item 'ISG+ memory CD4 T cell'\n",
       "\\item 'ISG+ memory CD8 T cell'\n",
       "\\item 'ISG+ naive B cell'\n",
       "\\item 'ISG+ naive CD4 T cell'\n",
       "\\item 'ISG+ naive CD8 T cell'\n",
       "\\item 'KLRB1+ memory CD4 Treg'\n",
       "\\item 'KLRB1+ memory CD8 Treg'\n",
       "\\item 'KLRF1- effector Vd1 gdT'\n",
       "\\item 'KLRF1- GZMB+ CD27- EM CD8 T cell'\n",
       "\\item 'KLRF1- GZMB+ CD27- memory CD4 T cell'\n",
       "\\item 'KLRF1+ effector Vd1 gdT'\n",
       "\\item 'KLRF1+ GZMB+ CD27- EM CD8 T cell'\n",
       "\\item 'Memory CD4 Treg'\n",
       "\\item 'Memory CD8 Treg'\n",
       "\\item 'Naive CD4 Treg'\n",
       "\\item 'Naive Vd1 gdT'\n",
       "\\item 'pDC'\n",
       "\\item 'Plasma cell'\n",
       "\\item 'Platelet'\n",
       "\\item 'Proliferating NK cell'\n",
       "\\item 'Proliferating T cell'\n",
       "\\item 'SOX4+ naive CD4 T cell'\n",
       "\\item 'SOX4+ naive CD8 T cell'\n",
       "\\item 'SOX4+ Vd1 gdT'\n",
       "\\item 'Transitional B cell'\n",
       "\\item 'Type 2 polarized memory B cell'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Activated memory B cell'\n",
       "2. 'Adaptive NK cell'\n",
       "3. 'ASDC'\n",
       "4. 'BaEoMaP cell'\n",
       "5. 'C1Q+ CD16 monocyte'\n",
       "6. 'CD14+ cDC2'\n",
       "7. 'CD27- effector B cell'\n",
       "8. 'CD27+ effector B cell'\n",
       "9. 'CD4 MAIT'\n",
       "10. 'CD56bright NK cell'\n",
       "11. 'CD8 MAIT'\n",
       "12. 'CD8aa'\n",
       "13. 'CD95 memory B cell'\n",
       "14. 'cDC1'\n",
       "15. 'CLP cell'\n",
       "16. 'CM CD4 T cell'\n",
       "17. 'CM CD8 T cell'\n",
       "18. 'CMP cell'\n",
       "19. 'Core CD14 monocyte'\n",
       "20. 'Core CD16 monocyte'\n",
       "21. 'Core memory B cell'\n",
       "22. 'Core naive B cell'\n",
       "23. 'Core naive CD4 T cell'\n",
       "24. 'Core naive CD8 T cell '\n",
       "25. 'DN T cell'\n",
       "26. 'Early memory B cell'\n",
       "27. 'Erythrocyte'\n",
       "28. 'GZMB- CD27- EM CD4 T cell'\n",
       "29. 'GZMB- CD27+ EM CD4 T cell'\n",
       "30. 'GZMB+ Vd2 gdT'\n",
       "31. 'GZMK- CD27+ EM CD8 T cell'\n",
       "32. 'GZMK- CD56dim NK cell'\n",
       "33. 'GZMK+ CD27+ EM CD8 T cell'\n",
       "34. 'GZMK+ CD56dim NK cell'\n",
       "35. 'GZMK+ memory CD4 Treg'\n",
       "36. 'GZMK+ Vd2 gdT'\n",
       "37. 'HLA-DRhi cDC2'\n",
       "38. 'IL1B+ CD14 monocyte'\n",
       "39. 'ILC'\n",
       "40. 'Intermediate monocyte'\n",
       "41. 'ISG+ CD14 monocyte'\n",
       "42. 'ISG+ CD16 monocyte'\n",
       "43. 'ISG+ CD56dim NK cell'\n",
       "44. 'ISG+ cDC2'\n",
       "45. 'ISG+ MAIT'\n",
       "46. 'ISG+ memory CD4 T cell'\n",
       "47. 'ISG+ memory CD8 T cell'\n",
       "48. 'ISG+ naive B cell'\n",
       "49. 'ISG+ naive CD4 T cell'\n",
       "50. 'ISG+ naive CD8 T cell'\n",
       "51. 'KLRB1+ memory CD4 Treg'\n",
       "52. 'KLRB1+ memory CD8 Treg'\n",
       "53. 'KLRF1- effector Vd1 gdT'\n",
       "54. 'KLRF1- GZMB+ CD27- EM CD8 T cell'\n",
       "55. 'KLRF1- GZMB+ CD27- memory CD4 T cell'\n",
       "56. 'KLRF1+ effector Vd1 gdT'\n",
       "57. 'KLRF1+ GZMB+ CD27- EM CD8 T cell'\n",
       "58. 'Memory CD4 Treg'\n",
       "59. 'Memory CD8 Treg'\n",
       "60. 'Naive CD4 Treg'\n",
       "61. 'Naive Vd1 gdT'\n",
       "62. 'pDC'\n",
       "63. 'Plasma cell'\n",
       "64. 'Platelet'\n",
       "65. 'Proliferating NK cell'\n",
       "66. 'Proliferating T cell'\n",
       "67. 'SOX4+ naive CD4 T cell'\n",
       "68. 'SOX4+ naive CD8 T cell'\n",
       "69. 'SOX4+ Vd1 gdT'\n",
       "70. 'Transitional B cell'\n",
       "71. 'Type 2 polarized memory B cell'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Activated memory B cell\"             \n",
       " [2] \"Adaptive NK cell\"                    \n",
       " [3] \"ASDC\"                                \n",
       " [4] \"BaEoMaP cell\"                        \n",
       " [5] \"C1Q+ CD16 monocyte\"                  \n",
       " [6] \"CD14+ cDC2\"                          \n",
       " [7] \"CD27- effector B cell\"               \n",
       " [8] \"CD27+ effector B cell\"               \n",
       " [9] \"CD4 MAIT\"                            \n",
       "[10] \"CD56bright NK cell\"                  \n",
       "[11] \"CD8 MAIT\"                            \n",
       "[12] \"CD8aa\"                               \n",
       "[13] \"CD95 memory B cell\"                  \n",
       "[14] \"cDC1\"                                \n",
       "[15] \"CLP cell\"                            \n",
       "[16] \"CM CD4 T cell\"                       \n",
       "[17] \"CM CD8 T cell\"                       \n",
       "[18] \"CMP cell\"                            \n",
       "[19] \"Core CD14 monocyte\"                  \n",
       "[20] \"Core CD16 monocyte\"                  \n",
       "[21] \"Core memory B cell\"                  \n",
       "[22] \"Core naive B cell\"                   \n",
       "[23] \"Core naive CD4 T cell\"               \n",
       "[24] \"Core naive CD8 T cell \"              \n",
       "[25] \"DN T cell\"                           \n",
       "[26] \"Early memory B cell\"                 \n",
       "[27] \"Erythrocyte\"                         \n",
       "[28] \"GZMB- CD27- EM CD4 T cell\"           \n",
       "[29] \"GZMB- CD27+ EM CD4 T cell\"           \n",
       "[30] \"GZMB+ Vd2 gdT\"                       \n",
       "[31] \"GZMK- CD27+ EM CD8 T cell\"           \n",
       "[32] \"GZMK- CD56dim NK cell\"               \n",
       "[33] \"GZMK+ CD27+ EM CD8 T cell\"           \n",
       "[34] \"GZMK+ CD56dim NK cell\"               \n",
       "[35] \"GZMK+ memory CD4 Treg\"               \n",
       "[36] \"GZMK+ Vd2 gdT\"                       \n",
       "[37] \"HLA-DRhi cDC2\"                       \n",
       "[38] \"IL1B+ CD14 monocyte\"                 \n",
       "[39] \"ILC\"                                 \n",
       "[40] \"Intermediate monocyte\"               \n",
       "[41] \"ISG+ CD14 monocyte\"                  \n",
       "[42] \"ISG+ CD16 monocyte\"                  \n",
       "[43] \"ISG+ CD56dim NK cell\"                \n",
       "[44] \"ISG+ cDC2\"                           \n",
       "[45] \"ISG+ MAIT\"                           \n",
       "[46] \"ISG+ memory CD4 T cell\"              \n",
       "[47] \"ISG+ memory CD8 T cell\"              \n",
       "[48] \"ISG+ naive B cell\"                   \n",
       "[49] \"ISG+ naive CD4 T cell\"               \n",
       "[50] \"ISG+ naive CD8 T cell\"               \n",
       "[51] \"KLRB1+ memory CD4 Treg\"              \n",
       "[52] \"KLRB1+ memory CD8 Treg\"              \n",
       "[53] \"KLRF1- effector Vd1 gdT\"             \n",
       "[54] \"KLRF1- GZMB+ CD27- EM CD8 T cell\"    \n",
       "[55] \"KLRF1- GZMB+ CD27- memory CD4 T cell\"\n",
       "[56] \"KLRF1+ effector Vd1 gdT\"             \n",
       "[57] \"KLRF1+ GZMB+ CD27- EM CD8 T cell\"    \n",
       "[58] \"Memory CD4 Treg\"                     \n",
       "[59] \"Memory CD8 Treg\"                     \n",
       "[60] \"Naive CD4 Treg\"                      \n",
       "[61] \"Naive Vd1 gdT\"                       \n",
       "[62] \"pDC\"                                 \n",
       "[63] \"Plasma cell\"                         \n",
       "[64] \"Platelet\"                            \n",
       "[65] \"Proliferating NK cell\"               \n",
       "[66] \"Proliferating T cell\"                \n",
       "[67] \"SOX4+ naive CD4 T cell\"              \n",
       "[68] \"SOX4+ naive CD8 T cell\"              \n",
       "[69] \"SOX4+ Vd1 gdT\"                       \n",
       "[70] \"Transitional B cell\"                 \n",
       "[71] \"Type 2 polarized memory B cell\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(master$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8a7d926a-d22e-4f2a-adc1-003a96e06ce9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# #individual testing\n",
    "# cell = \"CD8aa\"\n",
    "\n",
    "# cell_filt <- master %>% dplyr::filter(master$celltype == cell, padj <=0.1)\n",
    "\n",
    "# DEG_male<- cell_filt[cell_filt$padj <= 0.1 & cell_filt$type == \"DEGs_Male\", \"X\"]\n",
    "# #DEG_male\n",
    "# DEG_female <- cell_filt[cell_filt$padj <= 0.1 & cell_filt$type == \"DEGs_Female\", \"X\"]\n",
    "# #DEG_female\n",
    "\n",
    "#  # Remove NA values from DEGs\n",
    "# DEG_male <- na.omit(DEG_male)\n",
    "# DEG_female <- na.omit(DEG_female)\n",
    "\n",
    "\n",
    "# #Convert filtered DEGs to sets\n",
    "# set_DEG_male <- as.set(DEG_male)\n",
    "# set_DEG2_female <- as.set(DEG_female)\n",
    "# # set_DEG_male\n",
    "# # set_DEG_female\n",
    "\n",
    "# jaccard_sim <- jaccard(set_DEG_male, set_DEG2_female)  #<<< change JI function here\n",
    "# jaccard_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "edf939e9-6dad-46f0-a589-3b453e659c7f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               celltype jaccard_similarity total_DEGs_male\n",
      "1               Activated memory B cell                NaN               0\n",
      "2                      Adaptive NK cell                NaN               0\n",
      "3                                  ASDC                NaN               0\n",
      "4                          BaEoMaP cell                NaN               0\n",
      "5                    C1Q+ CD16 monocyte                NaN               0\n",
      "6                            CD14+ cDC2                NaN               0\n",
      "7                 CD27- effector B cell         0.00000000               0\n",
      "8                 CD27+ effector B cell         0.00000000               0\n",
      "9                              CD4 MAIT         0.00000000               1\n",
      "10                   CD56bright NK cell                NaN               0\n",
      "11                             CD8 MAIT         0.12500000               4\n",
      "12                                CD8aa         0.00000000               6\n",
      "13                   CD95 memory B cell                NaN               0\n",
      "14                                 cDC1                NaN               0\n",
      "15                             CLP cell                NaN               0\n",
      "16                        CM CD4 T cell         0.05882353               6\n",
      "17                        CM CD8 T cell         0.06779661              50\n",
      "18                             CMP cell                NaN               0\n",
      "19                   Core CD14 monocyte         0.00000000               2\n",
      "20                   Core CD16 monocyte         0.00000000               1\n",
      "21                   Core memory B cell         0.33333333               1\n",
      "22                    Core naive B cell                NaN               0\n",
      "23                Core naive CD4 T cell         0.16666667             172\n",
      "24               Core naive CD8 T cell          0.17241379              34\n",
      "25                            DN T cell                NaN               0\n",
      "26                  Early memory B cell         0.00000000               0\n",
      "27                          Erythrocyte         0.00000000               1\n",
      "28            GZMB- CD27- EM CD4 T cell         0.06250000               2\n",
      "29            GZMB- CD27+ EM CD4 T cell         0.05882353               7\n",
      "30                        GZMB+ Vd2 gdT         0.00000000               0\n",
      "31            GZMK- CD27+ EM CD8 T cell         0.00000000               2\n",
      "32                GZMK- CD56dim NK cell                NaN               0\n",
      "33            GZMK+ CD27+ EM CD8 T cell         0.00000000              11\n",
      "34                GZMK+ CD56dim NK cell                NaN               0\n",
      "35                GZMK+ memory CD4 Treg                NaN               0\n",
      "36                        GZMK+ Vd2 gdT                NaN               0\n",
      "37                        HLA-DRhi cDC2         0.00000000               0\n",
      "38                  IL1B+ CD14 monocyte         0.00000000               0\n",
      "39                                  ILC                NaN               0\n",
      "40                Intermediate monocyte         0.00000000               2\n",
      "41                   ISG+ CD14 monocyte         0.00000000               2\n",
      "42                   ISG+ CD16 monocyte         0.00000000               0\n",
      "43                 ISG+ CD56dim NK cell                NaN               0\n",
      "44                            ISG+ cDC2                NaN               0\n",
      "45                            ISG+ MAIT                NaN               0\n",
      "46               ISG+ memory CD4 T cell                NaN               0\n",
      "47               ISG+ memory CD8 T cell                NaN               0\n",
      "48                    ISG+ naive B cell                NaN               0\n",
      "49                ISG+ naive CD4 T cell         0.00000000               0\n",
      "50                ISG+ naive CD8 T cell                NaN               0\n",
      "51               KLRB1+ memory CD4 Treg                NaN               0\n",
      "52               KLRB1+ memory CD8 Treg                NaN               0\n",
      "53              KLRF1- effector Vd1 gdT                NaN               0\n",
      "54     KLRF1- GZMB+ CD27- EM CD8 T cell         0.00000000               0\n",
      "55 KLRF1- GZMB+ CD27- memory CD4 T cell         0.00000000               1\n",
      "56              KLRF1+ effector Vd1 gdT                NaN               0\n",
      "57     KLRF1+ GZMB+ CD27- EM CD8 T cell         0.00000000               4\n",
      "58                      Memory CD4 Treg                NaN               0\n",
      "59                      Memory CD8 Treg                NaN               0\n",
      "60                       Naive CD4 Treg         0.04395604              73\n",
      "61                        Naive Vd1 gdT                NaN               0\n",
      "62                                  pDC                NaN               0\n",
      "63                          Plasma cell         0.00000000               0\n",
      "64                             Platelet                NaN               0\n",
      "65                Proliferating NK cell                NaN               0\n",
      "66                 Proliferating T cell         0.00000000               4\n",
      "67               SOX4+ naive CD4 T cell                NaN               0\n",
      "68               SOX4+ naive CD8 T cell         0.00000000               1\n",
      "69                        SOX4+ Vd1 gdT                NaN               0\n",
      "70                  Transitional B cell                NaN               0\n",
      "71       Type 2 polarized memory B cell                NaN               0\n",
      "   total_DEGs_female common_DEGs\n",
      "1                  0           0\n",
      "2                  0           0\n",
      "3                  0           0\n",
      "4                  0           0\n",
      "5                  0           0\n",
      "6                  0           0\n",
      "7                  2           0\n",
      "8                  1           0\n",
      "9                  0           0\n",
      "10                 0           0\n",
      "11                 5           1\n",
      "12                10           0\n",
      "13                 0           0\n",
      "14                 0           0\n",
      "15                 0           0\n",
      "16                12           1\n",
      "17                13           4\n",
      "18                 0           0\n",
      "19                37           0\n",
      "20                 1           0\n",
      "21                 3           1\n",
      "22                 0           0\n",
      "23               178          50\n",
      "24                68          15\n",
      "25                 0           0\n",
      "26                 1           0\n",
      "27                 0           0\n",
      "28                15           1\n",
      "29                11           1\n",
      "30                 5           0\n",
      "31                 3           0\n",
      "32                 0           0\n",
      "33                 7           0\n",
      "34                 0           0\n",
      "35                 0           0\n",
      "36                 0           0\n",
      "37                 1           0\n",
      "38                13           0\n",
      "39                 0           0\n",
      "40                 4           0\n",
      "41                 4           0\n",
      "42                 2           0\n",
      "43                 0           0\n",
      "44                 0           0\n",
      "45                 0           0\n",
      "46                 0           0\n",
      "47                 0           0\n",
      "48                 0           0\n",
      "49                 9           0\n",
      "50                 0           0\n",
      "51                 0           0\n",
      "52                 0           0\n",
      "53                 0           0\n",
      "54                 2           0\n",
      "55                 0           0\n",
      "56                 0           0\n",
      "57                 0           0\n",
      "58                 0           0\n",
      "59                 0           0\n",
      "60                22           4\n",
      "61                 0           0\n",
      "62                 0           0\n",
      "63                 4           0\n",
      "64                 0           0\n",
      "65                 0           0\n",
      "66                 0           0\n",
      "67                 0           0\n",
      "68                 0           0\n",
      "69                 0           0\n",
      "70                 0           0\n",
      "71                 0           0\n"
     ]
    }
   ],
   "source": [
    "# library(sets)\n",
    "# library(dplyr)\n",
    "\n",
    "# Initialize an empty dataframe to store results\n",
    "results <- data.frame(celltype = character(),\n",
    "                      jaccard_similarity = numeric(),\n",
    "                       total_DEGs_male = integer(),\n",
    "                       total_DEGs_female = integer(),\n",
    "                       common_DEGs = integer(),\n",
    "                       stringsAsFactors = FALSE)\n",
    "\n",
    "# Iterate through cell types\n",
    "for (cell in unique(master$celltype)) {\n",
    "  \n",
    "    # Filter data for the current cell type\n",
    "  cell_filt <- master %>% dplyr::filter(celltype == cell, padj <= 0.05)\n",
    "  \n",
    "  # Filter DEGs for males and females\n",
    "  DEG_male <- cell_filt[cell_filt$padj <= 0.05 & cell_filt$type == \"DEGs_Male\", \"X\"]\n",
    "  DEG_female <- cell_filt[cell_filt$padj <= 0.05 & cell_filt$type == \"DEGs_Female\", \"X\"]\n",
    "  \n",
    "  # Remove NA values from DEGs\n",
    "  DEG_male <- na.omit(DEG_male)\n",
    "  DEG_female <- na.omit(DEG_female)\n",
    "  \n",
    "  # Convert filtered DEGs to sets\n",
    "  set_DEG_male <- as.set(DEG_male)\n",
    "  set_DEG_female <- as.set(DEG_female)\n",
    "  \n",
    "  # Calculate total DEGs counts\n",
    "  total_DEGs_male <- length(DEG_male)\n",
    "  total_DEGs_female <- length(DEG_female)\n",
    "  \n",
    "  # Calculate common DEGs counts\n",
    "  common_DEGs <- length(intersect(set_DEG_male, set_DEG_female))\n",
    "  \n",
    "  # Calculate Jaccard similarity\n",
    "  jaccard_sim <- jaccard(set_DEG_male, set_DEG_female)\n",
    "  \n",
    "  # Append results to dataframe\n",
    "  results <- rbind(results, data.frame(celltype = cell,\n",
    "                                       jaccard_similarity = jaccard_sim,\n",
    "                                        total_DEGs_male = total_DEGs_male,\n",
    "                                        total_DEGs_female = total_DEGs_female,\n",
    "                                        common_DEGs = common_DEGs))\n",
    "}\n",
    "\n",
    "# Print results\n",
    "print(results)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "99eeaf98-ab8e-4d0b-80e3-97e503cc968f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "write.csv(results, \"/home/jupyter/IH-A-Aging-Analysis-Notebooks_old/Mansi_Notebooks/scRNA/Figure_Notebooks/Intermediate_files/Jaccard_index_df_FvsM_each_celltype_padj0.05.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dbf1756-2b40-4335-93ce-876221457142",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4",
   "language": "R",
   "name": "ir4"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
